### DISCUSSION TITLE
USB Enumeration/SET_ADDRESS

### DISCUSSION THREAD

**Discussion by perlindgren** [hathach/tinyusb#1522] [Q&A] [answered]
When fiddling with the TinyUsb stack on the samv71 (xplained-ultra), it seems that before the address has been set there are other requests on the ctrl endpoint.

Looking at:

https://www.beyondlogic.org/usbnutshell/usb6.shtml
![image](https://user-images.githubusercontent.com/4089704/174352182-d5be1aa2-43cc-4379-bc99-492f459a8428.png)

After a EORST has been received, we were expecting a SET_ADDRESS transaction, but instead we (first received) a GET_DESCRIPTOR request. 

Regarding this I/we have a two questions.

1) Is it expected that the HOST would send a GET_DESCRIPTOR request. HOST is linux (same behavior under nix/arch).
2) How does TinyUsb treat such request (since it is does not yet have any address), is the request just ignored or, does TinyUsb answer the HOST somehow?

Thanks in advance.
/Per

**Comment by Ryzee119** [comment, marked as answer]
After the device is reset, the host will often send a get device descriptor request to address 0. This is to determine the max packet size of the control endpoint. This first request will just use a default max packet size (8bytes for LS, 64bytes for HS/FS normally)

It will then perform a set address, then all future control requests will be to the new address with the correct max packet size.

The exact steps of enumeration are not well defined and there is some flexibility. Generally you want a device to be at address 0 for as short time as possible but it certainly is common for a GET_DESCRIPTOR to be performed after device reset to address 0.

TinyUSB will/should answer to address zero after a device reset until a set address is received. So this shouldnt be a problem

Heres a blog post from a windows developer on the windows 7 USB enumeration process. You will see it does something similar:
https://techcommunity.microsoft.com/t5/microsoft-usb-blog/how-does-usb-stack-enumerate-a-device/ba-p/270685

**Comment by perlindgren** [comment]
Thanks a bunch, that clarifies the situation, much appreciated!

I was blindly looking at the datasheet for the atsam x7x, and was falsely led to believe a SETUP/SET_ADDRESS was the first transaction of the enumeration.

Best regards
  Per
